operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/GroupVersionKind.java

```diff
@@ -2,12 +2,15 @@
 
 import java.util.Objects;
 
+import io.fabric8.kubernetes.api.Pluralize;
 import io.fabric8.kubernetes.api.model.HasMetadata;
 
 public class GroupVersionKind {
   private final String group;
   private final String version;
   private final String kind;
+  private final String plural;
```

Comment:
What is the use case of plural?

Reply:
it is not present on go side either afaik.

Reply:
While the plural can most of the time be inferred from the kind, it could be useful to know the precise plural version associated with resources identified by this GVK, if it is known. In particular, this would be useful for the linked PR (see: https://github.com/quarkiverse/quarkus-operator-sdk/pull/931/files#diff-b780cb662fc5c0a824655ac29ced7dca9efd35a33f771cd3f3a84a46f21b7cebR97 in particular).
The plural form is used in many spots, this is actually what is used to identify the REST endpoint to access when dealing with resources so knowing its precise form (if known) could be useful for pretty much anything that deals with RBACs.
To be clear, this is a convenience method which should be transparent to people not using it.

Reply:
Sorry did not see the link. My only concern is, that in JOSDK scop this has no usage whatsoever, that might lead to a confusion. Currently it is an internal technical class with a very well defined scope.  So in that regard basically we have leaking requirements from QOSDK to JOSDK, shouldn't be rather a `GroupVersionKindPlural` class (extending GVK) created in QOSDK, so we have nice between the projects?!

Reply:
This isn't a requirement but this is something (what is the plural form?) that someone might want to know when given a GVK. It makes sense on its own, imo. We could rename the class but I think it makes sense as-is, not to mention that keeping the same name would avoid breaking the API and thus requiring waiting for a major version for this feature, while I would like to get that released in a patch release quickly if possible.

Reply:
I don't mean renaming the class, but creating a class in QOSDK that extends this. 
It is not used anywhere in JOSDK - I mean plural, it is not in controller runtime gvk either. It might be needed for some completely different purpose in QOSDK to have this information, but then it should be a representation for it in QOSDK not here.

Reply:
> Sorry did not see the link. My only concern is, that in JOSDK scope this has no usage whatsoever, that might lead to confusion. 

This wouldn't be the first thing like this, even for stuff that isn't used in QOSDK. That's a property of framework to provide things that are not used within the framework.

Reply:
Look this isn't nitpicking, this is a central technical class that I want to keep as clean as possible. GVK plays a important role in dynamic APIs, check the glue operator: https://github.com/csviri/kubernetes-glue-operator
basically all the keys in caches and stuff are built on this. This is reasonable to assume from anyone that builds dynamic stuff. So we really want to have this as clean as possible.  

It is very simple to create a `GroupVersionKindPlural` in QOSDK that simply extends `GroupVersionKind` and the boundaries will be very clean.

> This wouldn't be the first thing like this, even for stuff that isn't used in QOSDK. 

We usually make stuff open for extension, I think we should not leak back notions until is not necessary.

> That's a property of framework to provide things that are not used within the framework.

We tend to have at least Sample / Integrations Test for every feature. Plural is really not needed anywhere in JOSDK atm.

Reply:
> Look this isn't nitpicking, this is a central technical class that I want to keep as clean as possible. GVK plays a important role in dynamic APIs, check the glue operator: https://github.com/csviri/kubernetes-glue-operator basically all the keys in caches and stuff are built on this. This is reasonable to assume from anyone that builds dynamic stuff. So we really want to have this as clean as possible.

You make an argument against an extension to a class that could benefit other projects based on the fact that another project uses that class. That seems rather self-serving. There are no more reason to "help" the Glue operator by keeping GVK "clean" than helping other projects by adding the plural version here.
Moreover, how exactly is adding the plural form (which is central to how resources are handled by kubernetes) "dirtying" the GVK class? It's not like it's adding something that's completely unrelated: plural is derived from kind so it makes sense to have it there, especially when the derivation is not trivial and can't only be inferred from the kind. Classes that use GVK lose nothing by having GVK provide a plural form as well.

> 
> It is very simple to create a `GroupVersionKindPlural` in QOSDK that simply extends `GroupVersionKind` and the boundaries will be very clean.

That is not true because some information might be lost by not providing the plural form. It might actually make sense to make `getPlural` return `Optional<String>` instead to convey that information.

If anything, you could make a point that GVK shouldn't be exposed at all outside of JOSDK. You could also make a point that JOSDK should have used the GVK class from the Fabric8 client instead of creating its own. JOSDK created its own version and exposed it outside of it, that's OK and it's used by other projects, that's also OK. I'm arguing that this class could be made even more useful to outside projects by making the plural form available which JOSDK is often able to compute properly, which might not be the case for external projects that might lack the information JOSDK has.

The goal of a framework is to be useful to external projects, not adhere to some weird concept of "purity". Having plural on GVK increases integration opportunities by exposing information that JOSDK is uniquely positioned to compute correctly.

Reply:
What I wanted to point is that the scope of this class now is to identify a Kubernetes object type. That is done by using Group, Version, Kind in Kubernetes. Plural plays a role on completely different places. 

It is completely fine to use GVK for this purpose outside of the project. But by adding the `plural` has nothing to do with this function (thus identifying a type).  That is what I mean clean.  Now you want to add plural, because it fits for some  functionality related to RBAC generation in QOSDK. That completely breaks the purpose of this class.

Reply:
I'm fine with having `GroupVersionKindPlural` that extends `GroupVersionKind`  in JOSDK, saying that this is a class that helps in other places. But not by adding plural into `GroupVersionKind` directly.

Reply:
see in Kubernetes: **https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime/schema#GroupVersionKind**

Reply:
note also the other types in that package: https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime/schema#pkg-types
with different purpose.